### ISSUE TITLE
Installer succeeds but no file in PipelineComponents folder

### ISSUE THREAD

**Issue by wkuhn** [keif888/SSISMHash#5]
32 bit installer (typical install) succeeds but afterward the component doesn't appear in SSDT (VS2010).  Looked in PipelineComponents folder (C:\Program Files (x86)\Microsoft SQL Server\110\DTS\PipelineComponents) but didn't find the multiple hash dll there so I suspect that the installer is quietly failing (or installing somewhere else?).

Tried the custom install to confirm that the installation path is correct (i.e., as noted above) and it was.

**Comment by keif888** [comment]
It sounds like you have used the 32 bit installer on a 64 bit machine.

The files will probably be in C:\Program Files\Microsoft SQL Server\110\DTS\PipelineComponents as there is no x86 on a 32 bit machine.

Can you please try the 64 bit installer, which will place the files where required, after uninstalling the 32 bit version.

Keith Martin
_____________________________

**Comment by wkuhn** [comment]
Keith,

Thanks for the quick response.  I should also add thanks for the great hash
transform.  We've used it for a few years and it works like a champ.

I did as you suggested:
1.  Uninstall the 32 bit version
2.  Install the 64 bit version
3.  Look in C:\Program Files\Microsoft SQL
Server\110\DTS\PipelineComponents for the dll

However there is no dll in that directory.

I did not install all of the features of SQL server (only SQL Server data
tools which is a VS2010 shell and BI projects).

I had performed a fresh install of windows 7 months ago and the multiple
hash installed without a hitch at that time.  I'm only having difficulty on
my fresh install of windows 10.  Just sharing that in case you might know
of any issues.

What's odd is that the installer doesn't report any errors even though it
doesn't put the file in PipelineComponents.

Thanks for the support.

Bill

**Comment by keif888** [comment]
I will have to build a machine to that combination and test what the installer is doing.  Windows 10 professional 64 bit, SSDT for SQL Server 2012.

That won't be until this afternoon, it's 7:41am here.  In the mean time, would it be possible for you to uninstall, and reinstall from the command line with msiexec /package file.msi /log c:\somewhere\hashinstall.log with file.msi replaced with the file you downloaded and somewhere being a folder on your machine?

Keith Martin
________________________________

**Comment by keif888** [comment]
G'day,
I have reread your comment.
You have installed SQL 2012 tools only.

I built a clean windows 10 machine, and installed SQL 2012 SP1 Developer, selecting only SQL Server Data Tools.
I left all paths as default.

I did a typical install, using the 1.7.0.0 msi file as follows:

C:\Users\Win10\Downloads>msiexec /package MultipleHashInstallerx64.msi /log c:\users\Win10\Downloads\mhashinstall.log

C:\Users\Win10\Downloads>dir \multipl*.dll /s
 Volume in drive C has no label.
 Volume Serial Number is 9453-529D
Directory of C:\Program Files (x86)\Microsoft SQL Server\110\DTS\PipelineComponents
07/07/2017  05:59 PM           172,032 MultipleHash2012.dll
               1 File(s)        172,032 bytes
 Directory of C:\Windows\Microsoft.NET\assembly\GAC_MSIL\MultipleHash2012\v4.0_1.0.0.0__51c551904274ab44
11/07/2017  09:40 PM           172,032 MultipleHash2012.dll
               1 File(s)        172,032 bytes
     Total Files Listed:
               2 File(s)        344,064 bytes
               0 Dir(s)  53,382,836,224 bytes free


This is what should happen, and Multiple Hash shown up in the SSIS Toolbox.

**I have been unable to replicate your issue.**

If you could please check the following folders for the DLL?
C:\Program Files (x86)\Microsoft SQL Server\110\DTS\PipelineComponents
C:\Windows\Microsoft.NET\assembly\GAC_MSIL\MultipleHash2012

And if it is not there, please uninstall, and reinstall with the following command:
msiexec /package MultipleHashInstallerx64.msi /log c:\users\UserName\Downloads\mhashinstall.log

And upload the log file, so I can diagnose what is going wrong.

FYI:
I recommend updating to the SSDT 17.1 release, which is able to target SQL 2012, 2014, 2016 and 2017 from one development IDE.

**Comment by wkuhn** [comment]
Keith,

1.  Confirmed that the MultipleHash2012.dll is present
in C:\Windows\Microsoft.NET\assembly\GAC_MSIL\MultipleHash2012\v4.0_1.0.0.0__51c551904274ab44
from my x64 install yesterday.
2.  However, the dll is not present in C:\Program Files\Microsoft SQL
Server\110\DTS\PipelineComponents
3.  Uninstalled the x64 version.
4.  Attempted install of x32 version first from command line with log
option.
5.  Installer completes and confirmed dll is present in
C:\Windows\Microsoft.NET\assembly\GAC_MSIL\MultipleHash2012\v4.0_1.0.0.0__51c551904274ab44
but not in the PipelineComponents (ran your dir \multipl*.dll /s command
from c:)
6.  Uninstalled x32 version
7.  Repeated 4 and 5 for x86 with same result (file in gac_msil
subdirectory but not in PipelineComponents subdirectory)

Attached are the .log files for both the x32 and x64 installs.

I do plan on installing visual studio 2015 with the latest version of SSDT
in the near future.  I was actually going to install it yesterday but the
microsoft site (msdn/my.visualstudio.com) was throwing 500 errors so I
couldn't proceed on that.  In the meantime I was hoping to simply replicate
my previous development environment.

Thanks for the support.

Question:  Can I simply copy the dll (appropriate bit version, of course)
to the PipelineComponents directory or is there something else that the
installer is trying to do?

Bill

**Comment by wkuhn** [comment]
Keith,

Okay, I worked around the problem but I'm not sure of the cause of the
problem.

I tried to install using a local admin account rather than my domain user
account (which is a member of the admin group).

When I installed as local admin the familiar "unknown publisher"
warning/prompt was displayed whereas when installing using my domain user
account it was not.

When I installed using the x32 installer the file appeared in the x86
PipelineComponents folder.  All good.  However, when I installed using the
x64 installer I didn't find the file in the 64 bit PipelineComponents
folder.

I reran the x64 installer using custom option and saw that it appears to be
installing to the x86 PipelineComponents folder.

Question:  Shouldn't each bit version of the installer install only to the
corresponding bit version PipelineComponents folder?  Aren't the dll's
different for each bit version?  I'm trying to avoid having to force 32 bit
execution in the designer and from sql server agent jobs.

Bill

**Comment by keif888** [comment]
Glad you have found a workaround.
Sounds like there is a domain level user policy in place that is preventing the installation working correctly.  I'll see if I can replicate that locally, so I can address it for others.

In answer to your question:

The installer detects what components of SQL Server are in place and only installs the dll and xml files to the appropriate folders.

- If you only have BIDS or SSDT's installed on a 32 bit operating system, then only the normal program files folder exists, and as such will be populated.  Of course only the 32 bit installer will work
- If you have a 32 bit installation of SQL Server on a 32 bit operating system, then there are only base program files folders, and they will be populated.  Of course only the 32 bit installer will work
- If you only have BIDS or SSDT's installed on a 64 bit operating system, then only the x86 folders will be populated.
- If you have a 32 bit installation of SQL Server on a 64 bit operating system, then only the x86 folders will be populated, regardless of which installer you use.
- If you have a 64 bit installation of SQL Server, then the x86 and normal program files folders will be populated.  You have to use the 64 bit installer for this.

There is only one version of the  DLL file for each edition of SQL Server.  The DLL is compiled to run either x86 or x64.

FYI:
For SQL 2008, SQL 2008 R2 and SQL 2012 the installer places the DLL into the GAC and the appropriate PipelineComponents folders.  It also installs an upgrade mapping XML file to UpgradeMappings.
For SQL 2014, SQL 2016 and in the future SQL 2017 the installer places the DLL into the GAC and the appropriate PipelineComponents folders.  It also installs an upgrade mapping and extension mapping XML file to UpgradeMappings.

**Comment by keif888** [comment]
I have pushed a new release out today, which should address this issue.
The underlying cause is that the installer wasn't setting it's properties to Secure, which means that if the account used to do the install is not a local administrator (membership of the administrators group via a domain account isn't quite enough), and you are not running msiexec from an elevated command prompt, then when the execution of the installation happens (after selecting typical/customer/complete) the directories detected/entered are not passed, and it installs to the drive letter that the installer is on, plus SSIS32_2012_PATH.
That is why the workaround worked, as the account was a local administrator.

**Comment by wkuhn** [comment]
Keith,

Works like a charm.

When I run the installer for the first time I get a warning from windows
that allows me to choose to either run or not run.  When I choose run
anyway I observe the MultipleHash2012.dll file appear in the
PipelineComponents folder.

I ran the 32 bit installer with success.  Then I ran the uninstall for the
32 bit and then ran the 64 bit installer with success.

Problem solved!

Thanks so much for the support.

I'd buy you a beer if you weren't on the other side of the world. :-)

Bill
